lib/CL/devices/remote/communication.c

```diff
@@ -570,16 +574,38 @@ pocl_network_connect (remote_server_data_t *data, int *fd, unsigned port,
   memcpy (&server, ai->ai_addr, ai->ai_addrlen);
   addrlen = ai->ai_addrlen;
 
+#ifdef ENABLE_VSOCK
   POCL_RETURN_ERROR_ON (
-      ((socket_fd = socket (ai->ai_family, ai->ai_socktype, IPPROTO_TCP)) == -1),
+      ((socket_fd = socket (ai->ai_family, ai->ai_socktype,
+                            ai->ai_family == AF_VSOCK ? 0 : IPPROTO_TCP))
+       == -1),
       CL_INVALID_DEVICE, "socket() returned errno: %i\n", errno);
+
+  if (ai->ai_family == AF_VSOCK)
+    {
+      host_freeaddrinfo (ai);
+    }
+  else
+    {
+      freeaddrinfo (ai);
+      err = pocl_remote_client_set_socket_options (socket_fd, bufsize,
```

Comment:
VSOCK's implementation for level SOL_SOCKET is the same as other sockets. The latest kernel additionally supports SO_ZEROCOPY, and for level == AF_VSOCK, it also supports several:

SO_VM_SOCKETS_BUFFER_SIZE
SO_VM_SOCKETS_BUFFER_MAX_SIZE
SO_VM_SOCKETS_BUFFER_MIN_SIZE
SO_VM_SOCKETS_CONNECT_TIMEOUT_NEW
SO_VM_SOCKETS_CONNECT_TIMEOUT_OLD
We can clearly see this logic in the code here:

https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/net/vmw_vsock/af_vsock.c?id=e27090b1413ff236ca1aec26d6b022149115de2c#n1642

The kernel directory linux/tools/testing/vsock has some examples of calling setsockopt, so my current idea is to add an address family in the current version to distinguish between IPPROTO_TCP and AF_VSOCK, and then further adjust these options in the future, such as:

You think SO_RCVBUF/SO_SNDBUF is not wise in VSOCK
Is it also a problem that NETWORK_BUF_SIZE_FAST is fixed to 4K for systems with a page size of 64k?

Reply:
Thanks for taking the time to look into this!
Yes, making the exact options being set conditional on the address family is exactly what I was getting at.

I don't have any concrete knowledge on the SO_RCVBUF/SO_SNDBUF bit, but intuitively I would assume that if AF_VSOCK works by simply sharing memory between processes, it is likely most efficient if each buffer is an exact multiple of the memory page size. Looking at the code it looks like it defaults to 256KiB so maybe it's best to simply leave that entirely untouched for now? We can always adjust it if someone takes the time to benchmark that and finds that other values work better.

SO_ZEROCOPY sounds like it could be useful, since AF_VSOCK essentially just shuffles memory around between processes. https://lwn.net/Articles/913969/ seems to suggest that at least for the "slow" socket (buffer data bulk transfers) it would be beneficial.

Reply:
The next stage will involve performance research on pocl vsock. We will submit a pull request once the research is completed.